src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,936 @@
+template <typename RandIt>
+RandIt filterPackedPathsByCellSharingHeuristic(const WeightedViaNodePackedPath &path,
+                                               const Partition &partition,
+                                               RandIt first,
+                                               RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNodePackedPath>();
+
+    const auto over_sharing_limit = [&](const auto &packed) {
+        const auto sharing = normalizedPackedPathSharingHeuristic(partition, path, packed);
+        return sharing > kAtLeastDifferentBy;
+    };
+
+    return std::remove_if(first, last, over_sharing_limit);
+}
+
+// Filters packed paths based on local optimality. Mutates range in-place.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterPackedPathsByLocalOptimalityHeuristic(const WeightedViaNodePackedPath &path,
+                                                   const Heap &forward_heap,
+                                                   const Heap &reverse_heap,
+                                                   RandIt first,
+                                                   RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNodePackedPath>();
+
+    // Check sub-path optimality on alternative path crossing the via node candidate.
+    //
+    //   s - - - v - - - t  our packed path made up of (from, to) edges
+    //        |--|--|       sub-paths "left" and "right" of v based on threshold
+    //        f  v  l       nodes in [v,f] and in [v, l] have to match predecessor in heaps
+
+    // Todo: this approach is efficient but works on packed paths only. Do we need to do a
+    // thorough check on the unpacked paths instead? Or do we even need to introduce two
+    // new thread-local heaps for the mld SearchEngineData and do proper s-t routing here?
+
+    BOOST_ASSERT(path.via.weight != INVALID_EDGE_WEIGHT);
+
+    // node == parent_in_main_heap(parent_in_side_heap(v)) -> plateaux at `node`
+    const auto has_plateaux_at_node = [&](
+        auto const node, const auto &main_heap, const auto &side_heap) {
+        BOOST_ASSERT(main_heap.WasInserted(node));
+        auto const parent_node = main_heap.GetData(node).parent;
+        return side_heap.WasInserted(parent_node) && side_heap.GetData(parent_node).parent == node;
+    };
+
+    // a plateaux is defined as a segment in which the search tree from s and the search tree from t
+    // overlap. An edge is part of such a plateaux around `v` if
+    // v == parent_in_reverse_search(parent_in_forward_search(v)).
+    // Here we calculate the last node no the plateaux in either direction
+    const auto plateaux_end = [&](auto node, auto const &main_heap, auto const &side_heap) {
```

Comment:
What is the meaning of `main_heap` and `side_heap` here? Is this the `forward_heap` and `backward_head`?